tests/ptf_runner.py

```diff
@@ -46,29 +50,90 @@ def get_dut_type(host):
+    if the image is of type 'mixed' or 'py3only'
+    """
+    pyvenv = host.stat(path="/root/env-python3/pyvenv.cfg")
+    if pyvenv["stat"]["exists"]:
+        return "mixed"
+    return "py3only"
+
+
+def get_test_path(testdir, testname):
+    """
+    Returns two values
+    - first: the complete path of the test based on testdir and testname.
+    - second: True if file is in 'py3' False otherwise
+    Raises FileNotFoundError if file is not found
+    """
+    curr_path = os.path.dirname(os.path.abspath(__file__))
+    base_path = pathlib.Path(curr_path).joinpath('..').joinpath('ansible/roles/test/files').joinpath(testdir)
+    idx = testname.find('.')
+    test_fname = testname + '.py' if idx == -1 else testname[:idx] + '.py'
+    chk_path = base_path.joinpath('py3').joinpath(test_fname)
+    if chk_path.exists():
+        return chk_path, True
+    chk_path = base_path.joinpath(test_fname)
+    if chk_path.exists():
+        return chk_path, False
+    raise FileNotFoundError("Testdir: {} Testname: {} File: {} not found".format(testdir, testname, chk_path))
+
+
+def is_py3_compat(test_fpath):
+    """
+    Returns True if the test can be run in a Python 3 environment
+    False otherwise.
+    """
+    if six.PY2:
+        raise Exception("must run in a Python 3 runtime")
+    with open(test_fpath, 'rb') as f:
+        code = f.read()
+        try:
+            ast.parse(code)
+        except SyntaxError:
+            return False
+        return True
+    # shouldn't get here
+    return False
+
+
 def ptf_runner(host, testdir, testname, platform_dir=None, params={},
                platform="remote", qlen=0, relax=True, debug_level="info",
                socket_recv_size=None, log_file=None, device_sockets=[], timeout=0, custom_options="",
-               module_ignore_errors=False, is_python3=False, async_mode=False, pdb=False):
-    # Call virtual env ptf for migrated py3 scripts.
-    # ptf will load all scripts under ptftests, it will throw error for py2 scripts.
-    # So move migrated scripts to seperated py3 folder avoid impacting py2 scripts.
+               module_ignore_errors=False, is_python3=None, async_mode=False, pdb=False):
+
     dut_type = get_dut_type(host)
     if dut_type == "kvm" and params.get("kvm_support", True) is False:
         logger.info("Skip test case {} for not support on KVM DUT".format(testname))
         return True
 
-    if is_python3:
-        path_exists = host.stat(path="/root/env-python3/bin/ptf")
-        if path_exists["stat"]["exists"]:
-            cmd = "/root/env-python3/bin/ptf --test-dir {} {}".format(testdir + '/py3', testname)
+    cmd = ""
+    ptf_img_type = get_ptf_image_type(host)
+    logger.info('PTF image type: {}'.format(ptf_img_type))
+    test_fpath, in_py3 = get_test_path(testdir, testname)
+    logger.info('Test file path {}, in py3: {}'.format(test_fpath, in_py3))
+    is_python3 = is_py3_compat(test_fpath)
+
+    ptf_cmd = None
```

Comment:
Probably add some comments here to better explain the logic here.
The expectation of PTF image is either "mixed" or "py3only". "mixed" means that global python is python2. Python3 and dependent package is installed in `/root/env-python3`. "py3only" means that there is no python2. The global python is python3 and it has all required dependent packages. 
For "mixed" ptf image, use python3 version `ptf` command to run python3 compatible ptf scripts. Use python2 version `ptf` command to run legacy python2 PTF scripts.
For "py3only" ptf image, always use the global python3 `ptf` command to run PTF scripts. Raise exception if ptf script is not python3 compatible.

Reply:
I have added the comment.